File: src/client/java/fr/hardel/asset_editor/client/javafx/routes/debug/DebugLogsPage.java
package fr.hardel.asset_editor.client.javafx.routes.debug;

import fr.hardel.asset_editor.client.javafx.VoxelColors;
import fr.hardel.asset_editor.client.javafx.VoxelFonts;
import fr.hardel.asset_editor.client.javafx.components.ui.Button;
import fr.hardel.asset_editor.client.javafx.components.ui.DataTable;
import fr.hardel.asset_editor.client.javafx.lib.Page;
import fr.hardel.asset_editor.client.javafx.lib.debug.DebugLogStore;
import fr.hardel.asset_editor.client.javafx.lib.debug.DebugLogStore.Entry;
import javafx.application.Platform;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import javafx.scene.layout.Region;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import net.minecraft.client.resources.language.I18n;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Map;

public final class DebugLogsPage extends VBox implements Page {

    private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("HH:mm:ss.SSS").withZone(ZoneId.systemDefault());

    private static final Color INFO_COLOR = Color.web("#38bdf8");
    private static final Color WARN_COLOR = Color.web("#fbbf24");
    private static final Color ERROR_COLOR = Color.web("#f87171");
    private static final Color SUCCESS_COLOR = Color.web("#34d399");

    private final DataTable<Entry> table = new DataTable<>();
    private final Label countLabel = new Label();
    private Runnable subscription;

    public DebugLogsPage() {
        getStyleClass().add("debug-log-page");
        setFillWidth(true);

        table.addColumn(I18n.get("debug:logs.column.time"), 100, this::timeCell);
        table.addColumn(I18n.get("debug:logs.column.level"), 80, this::levelCell);
        table.addColumn(I18n.get("debug:logs.column.category"), 100, this::categoryCell);
        table.addColumn(I18n.get("debug:logs.column.message"), -1, this::messageCell);
        table.setExpandFactory(this::detailNode);
        table.setIdExtractor(Entry::id);
        table.setPlaceholder(I18n.get("debug:logs.placeholder"));
        VBox.setVgrow(table, Priority.ALWAYS);

        countLabel.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 12));
        countLabel.setTextFill(VoxelColors.ZINC_500);

        Button clearBtn = new Button(Button.Variant.GHOST_BORDER, Button.Size.SM, I18n.get("debug:action.clear"));
        clearBtn.setOnAction(DebugLogStore::clear);

        Region spacer = new Region();
        HBox.setHgrow(spacer, Priority.ALWAYS);

        HBox toolbar = new HBox(8, countLabel, spacer, clearBtn);
        toolbar.setAlignment(Pos.CENTER_LEFT);
        toolbar.setPadding(new Insets(12, 16, 12, 16));

        getChildren().addAll(toolbar, table);

        sceneProperty().addListener((obs, oldScene, newScene) -> {
            if (newScene == null) {
                if (subscription != null) {
                    subscription.run();
                    subscription = null;
                }
                return;
            }
            if (subscription == null)
                subscription = DebugLogStore.subscribe(() -> Platform.runLater(this::refresh));
            refresh();
        });

        refresh();
    }

    private void refresh() {
        var entries = DebugLogStore.entries().stream()
            .filter(entry -> entry.category() != DebugLogStore.Category.NETWORK)
            .toList();
        countLabel.setText(I18n.get("debug:logs.count", entries.size()));
        table.setItems(entries);
    }

    private Label timeCell(Entry entry) {
        Label label = new Label(TIME_FORMAT.format(Instant.ofEpochMilli(entry.timestamp())));
        label.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 11));
        label.setTextFill(VoxelColors.ZINC_500);
        return label;
    }

    private Label levelCell(Entry entry) {
        Label label = new Label(I18n.get("debug:logs.level." + entry.level().name().toLowerCase()));
        label.setFont(VoxelFonts.of(VoxelFonts.Variant.SEMI_BOLD, 11));
        label.setTextFill(levelColor(entry.level()));
        return label;
    }

    private Label categoryCell(Entry entry) {
        Label label = new Label(I18n.get("debug:logs.category." + entry.category().name().toLowerCase()));
        label.setFont(VoxelFonts.of(VoxelFonts.Variant.MEDIUM, 11));
        label.setTextFill(VoxelColors.ZINC_400);
        return label;
    }

    private Label messageCell(Entry entry) {
        Label label = new Label(entry.message());
        label.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 13));
        label.setTextFill(VoxelColors.ZINC_300);
        label.setWrapText(true);
        return label;
    }

    private Node detailNode(Entry entry) {
        if (entry.data() == null || entry.data().isEmpty()) {
            Label empty = new Label(I18n.get("debug:logs.no_additional_data"));
            empty.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 12));
            empty.setTextFill(VoxelColors.ZINC_500);
            return empty;
        }

        VBox box = new VBox(4);
        for (Map.Entry<String, String> kv : entry.data().entrySet()) {
            Label line = new Label(kv.getKey() + ": " + kv.getValue());
            line.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 12));
            line.setTextFill(VoxelColors.ZINC_400);
            line.setWrapText(true);
            box.getChildren().add(line);
        }
        return box;
    }

    private static Color levelColor(DebugLogStore.Level level) {
        return switch (level) {
            case INFO -> INFO_COLOR;
            case WARN -> WARN_COLOR;
            case ERROR -> ERROR_COLOR;
            case SUCCESS -> SUCCESS_COLOR;
        };
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/routes/debug/DebugLogsPage.kt
package fr.hardel.asset_editor.client.compose.routes.debug

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import fr.hardel.asset_editor.client.compose.StudioColors
import fr.hardel.asset_editor.client.compose.StudioTypography
import fr.hardel.asset_editor.client.compose.components.page.debug.DEBUG_TIME_FORMAT
import fr.hardel.asset_editor.client.compose.components.page.debug.DebugLogsActionBar
import fr.hardel.asset_editor.client.compose.components.page.debug.DebugLogsExpandRow
import fr.hardel.asset_editor.client.compose.components.page.debug.copyDebugLogsToClipboard
import fr.hardel.asset_editor.client.compose.components.page.debug.debugLogLevelColor
import fr.hardel.asset_editor.client.compose.components.ui.datatable.DataTable
import fr.hardel.asset_editor.client.compose.components.ui.datatable.TableColumn
import fr.hardel.asset_editor.client.compose.lib.StudioContext
import fr.hardel.asset_editor.client.compose.lib.asFlow
import net.minecraft.client.resources.language.I18n
import java.time.Instant

@Composable
fun DebugLogsPage(context: StudioContext) {
    var currentPage by remember { mutableStateOf(0) }
    var expandedIds by remember { mutableStateOf(emptySet<Long>()) }
    val flow = remember(context) { context.debugMemory().logs().asFlow() }
    val snapshot by flow.collectAsState(context.debugMemory().logs().snapshot())
    val entries = snapshot.entries

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp)
    ) {
        DebugLogsActionBar(
            entryCount = entries.size,
            onCopyAll = { copyDebugLogsToClipboard(entries) },
            onClear = context.debugMemory().logs()::clear
        )

        DataTable(
            items = entries,
            columns = listOf(
                TableColumn(I18n.get("debug:logs.column.time"), weight = 0.8f) { entry ->
                    Text(
                        DEBUG_TIME_FORMAT.format(Instant.ofEpochMilli(entry.timestamp())),
                        style = StudioTypography.regular(11),
                        color = StudioColors.Zinc500
                    )
                },
                TableColumn(I18n.get("debug:logs.column.level"), weight = 0.7f) { entry ->
                    Text(
                        I18n.get("debug:logs.level.${entry.level().name.lowercase()}"),
                        style = StudioTypography.semiBold(11),
                        color = debugLogLevelColor(entry.level())
                    )
                },
                TableColumn(I18n.get("debug:logs.column.category"), weight = 0.9f) { entry ->
                    Text(
                        I18n.get("debug:logs.category.${entry.category().name.lowercase()}"),
                        style = StudioTypography.medium(11),
                        color = StudioColors.Zinc400
                    )
                },
                TableColumn(I18n.get("debug:logs.column.message"), weight = 2.6f) { entry ->
                    Text(entry.message(), style = StudioTypography.regular(13), color = StudioColors.Zinc300)
                }
            ),
            pageSize = 50,
            currentPage = currentPage,
            onPageChange = { page -> currentPage = page },
            placeholder = I18n.get("debug:logs.placeholder"),
            idExtractor = { entry -> entry.id() },
            expandedIds = expandedIds,
            onToggleExpand = { id -> expandedIds = if (id in expandedIds) expandedIds - id else expandedIds + id },
            expandContent = { entry -> DebugLogsExpandRow(entry) },
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 12.dp)
        )
    }
}
